GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs

```diff
@@ -0,0 +1,89 @@
+        [SerializeField] private GameObject _pageImagePosition;
+        [SerializeField] private float _sentenceSpeed;
+        [SerializeField] private Button _option;
+        [SerializeField] private GameObject[] _buttonPositions;
+        [SerializeField] private Image _dialogueBackground;
+        [SerializeField] private GameObject _dialogueBox;
+        private List<Button> _responseOptions = new List<Button>();
+        private Image previousImage;
+
+        private void RenderPageText(string pageName, string pageText)
+        {
+            _pageName.text = pageName;
+            _pageText.text = pageText;
+        }
+
+        private IEnumerator Play(Dialogue npc, Message npcMessage)
+        {
+            var sb = new StringBuilder();
+            var letters = npcMessage.MessageText.ToCharArray();
+            foreach (var letter in letters)
+            {
+                sb.Append(letter);
+                RenderPageText(npc.NpcName, sb.ToString());
+                yield return new WaitForSeconds(_sentenceSpeed);
+            }
+            yield return null;
+        }
+
+        public void PlayParagraphCycle(Dialogue npcDialogue, NpcImages npcImages, int paragraphNumber)
+        {
+            if (_dialogueBox.activeSelf == false) _dialogueBox.SetActive(true);
+            
+            if (paragraphNumber < 0)
+            {
+                EndDialogue();
+            }
+            else
+            {
+                if (previousImage != null) Destroy(previousImage.gameObject);
+                var newImage = Instantiate(npcImages.NpcImage[npcDialogue.NpcId].NpcMoodImages[npcDialogue.Messages[paragraphNumber].NpcMoodId], _pageImagePosition.transform);
+                newImage.transform.SetParent(_dialogueBackground.transform);
+                previousImage = newImage;
+                StartCoroutine(Play(npcDialogue, npcDialogue.Messages[paragraphNumber]));
+                GetResponse(npcDialogue, npcDialogue.Messages[paragraphNumber], npcImages);
+            }
+        }
+
+        private void GetResponse(Dialogue npcMessage, Message npcResponses, NpcImages npcImages)
+        {
+            foreach (Button buttonObject in _responseOptions)
+            {
+                Destroy(buttonObject.gameObject);
+            }
+            _responseOptions.Clear();
+            int buttonCount = 0;
+            foreach (var response in npcResponses.Responses)
+            {
+                
+                var button = Instantiate(_option, _buttonPositions[buttonCount].transform.position, _buttonPositions[buttonCount].transform.rotation);
+                button.transform.SetParent(_dialogueBackground.transform);
+                button.GetComponentInChildren<Text>().text = response.Reply;
+                _responseOptions.Add(button);
+                button.onClick.AddListener(() => PlayParagraphCycle(npcMessage, npcImages, response.Next));
+                ++buttonCount;
```

Comment:
why not just use a `for` loop here instead?

Reply:
Personal preference to use foreach like this, I find it clearer and nicer to use

Reply:
I advise changing it still. This is precisely what `for` was meant for - not to mention you are introducing additional overhead by relying on `foreach`, albeit a tiny amount, its slightly noisy code.